Fixes 11008 - Changed the view for discovered hosts show page.

alongoldboim opened the pull request:
I have also hidden the rest of the facts and added a button to expand the facts table.

alongoldboim commented:
@ohadlevy I'm not sure i have added all the facts you've wanted. need to check if i understood correctly.

ohadlevy commented:
@alongoldboim screenshot always help :)

ohadlevy commented on app/assets/javascripts/discovered_hosts.js:
```diff
@@ -0,0 +1,9 @@
+$(document).ready(function(){
+
+$('#moreFactsBtn').click(function () {
+    $(this).text(function(i, text){
```
indentation.. should be 2 spaces

ohadlevy commented on app/assets/javascripts/discovered_hosts.js:
```diff
@@ -0,0 +1,9 @@
+
+$('#moreFactsBtn').click(function () {
+    $(this).text(function(i, text){
+        return text === "More facts" ? "Less facts" : "More facts";
```
strings are not extract

ohadlevy commented on app/helpers/discovered_hosts_helper.rb:
```diff
@@ -44,4 +44,19 @@ def discovery_attribute(host, attr, default_value = _('N/A'))
+  def facts_table(values,with_header)
+    content_tag :table, class: "table table-bordered table-striped table-conden" do
+      rows = []
+      rows << content_tag(:th, _('Fact')) if with_header
```
a bit cryptic imho to write the whole table as ruby. if you are creating a table helper you could do something like

``` erb
<% table do %>
<thead>
<th>
..
</thead>
<tbody>
<tr>
...
</tr>
</tbody>
<% end %>
```

alongoldboim commented:
![selection_001](https://cloud.githubusercontent.com/assets/11405684/8801283/1c2eaa1e-2fc2-11e5-827a-a62cdd1b7f56.png)
![selection_002](https://cloud.githubusercontent.com/assets/11405684/8801284/1c30a850-2fc2-11e5-9db9-5bbffec7342b.png)

lzap commented on app/assets/javascripts/discovered_hosts.js:
```diff
@@ -0,0 +1,9 @@
+    })
+});
+
+});
```
Missing newline

lzap commented on lib/foreman_discovery/engine.rb:
```diff
@@ -19,6 +19,19 @@ class Engine < ::Rails::Engine
+              'discovered_hosts.js',
+          ]}}
+    end
+
```
Isn't this an overkill for such a tiny feature? I mean, this needs a change in our packaging. Can you make a change for both Red Hat and Debian packages? https://github.com/theforeman/foreman-packaging

lzap replied:
I think this could be also inline JavaScript, it's just one statement.

lzap replied:
Maybe instead of changing the button value, we can just hide it?

alongoldboim replied:
@lzap i can do it inline so we won't have to do a repackaging.
About the feature as a whole i haven't finished yet, this was just the proto.

alongoldboim committed:
Fixes 11008 - Changed the view for discovered hosts show page.

alongoldboim commented:
@lzap @ohadlevy does this look better? any advises?
![selection_004](https://cloud.githubusercontent.com/assets/11405684/8905317/f2c51e82-346d-11e5-8ef5-48207a11c02c.png)
![selection_005](https://cloud.githubusercontent.com/assets/11405684/8905318/f2c9c4a0-346d-11e5-9cbc-d3c7dacabbf5.png)

lzap commented:
As a prototype, this looks nice indeed.

alongoldboim commented:
Uploaded this version, if you could play around with it and give me some pointers ill be more then happy.

ohadlevy commented:
@alongoldboim do you think the general info is the most interesting information to show by default?

what about facts that do not fit a category?

alongoldboim commented:
@ohadlevy its not the most interesting your right, i got caught up with the whole category splitting :), ill gather the good facts in the general, the rest of the facts are in "miscellaneous" i added it on the commit but its not in the ss.
As for the look of it would you recommend on doing it different?

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -25,6 +25,9 @@ def index
+    @general= { :Highlights => ["interface", "ipaddress", "virtual", "manufacturer", "architecture", "memorysize_mb", "os", "macaddress"]}
+    @highlights ={:Network=>["discovery", "dhcp", "fqdn", "hostname", "link", "mtu", "net"], :Storage => ["blockdevice"], :Hardware => ["hardw", "manufacturer", "memo", "process"], :Software => ["bios", "os" ], :Etc => [] }
+    @facts = @host.facts_hash
     @range = nil
```
Things I'd scratch from general pane:
- memorysize_mb: 1855.26
- virtual: physical
- is_virtual: false
- macaddress: 52:54:00:26:fd:82

Things I miss in the general pane:

processorcount: 1
physicalprocessorcount: 1
memorytotal: 1.81 GB (this is more readable)
discovery_bootif: 52:54:00:13:05:13
discovery_bootip: 192.168.122.68

and total storage size as well.

alongoldboim replied:
i agree, but i still think ill leave the mac address though.

lzap replied:
Oh that was a pastebin error, scratch the mac of course.

lzap commented:
It looks very nice, several points:
- sort the facts by name in all panes
- rename "Etc" to somethign like "Miscellaneous"
- move discovery_version/release to "Software" pane
- since you know all interfaces names ("interfaces" fact), move all "*_inteface_name" facts in Network pane as well
- implement "..." or "Expand" for longer facts than 100 characters (typically ssh keys, an inline javascript one-way only is fine)
- move "productname" to general pane as well
- the most logical order of highlights is: Hardware, Storage, Network, Software, Etc
- open with Hardware expanded by default to leverage full space on the right side

Great addition, I like it.

lzap commented:
@ohadlevy fyi the pictures above were only UI protypes, it looks different in the code now...

alongoldboim commented:
@lzap thanks for the comments, applying changes.

lzap commented:
Ok, I don't see new commit yet. FYI unit tests failed.

alongoldboim commented:
Yea i need to see whats failing.
Monday it will be ready.

2015-07-30 11:34 GMT+03:00 Lukáš Zapletal notifications@github.com:

> Ok, I don't see new commit yet. FYI unit tests failed.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/theforeman/foreman_discovery/pull/200#issuecomment-126225202
> .

alongoldboim commented:
@lzap  I have added what you requested, check it and let me know if anything else is needed, thanks.

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -25,8 +25,15 @@ def index
+    @general =    {:Highlights => ["size", "productname", "memorysize", "interface", "manufacturer", "architecture", "os", "macaddress", "processorcount", "physicalprocessorcount", "discovery_boot"]}
+    @highlights = {:Hardware => ["hardw", "manufacturer", "memo", "process"],
+                  :Storage => ["blockdevice"],
+                  :Network=>["interfaces", "dhcp", "fqdn", "hostname", "link", "mtu", "net"] + @host.facts_hash["interfaces"].split(","),
```
Missing line space: `=>`

lzap commented on app/views/discovered_hosts/show.html.erb:
```diff
@@ -2,15 +2,46 @@
+          </div>
+          <div id="<%=keys.to_s+"panel"%>" class="panel-collapse collapse <%="in" if keys.to_s.include?"Hardware"%>" role="tabpanel">
+            <div class="panel-body">
+              <%@facts.sort_by { |k,v| k.downcase} %>
```
Missing space.

lzap commented on app/views/discovered_hosts/show.html.erb:
```diff
@@ -2,15 +2,46 @@
+            <h4 class="panel-title">
+              <a role="button" data-toggle="collapse" data-parent="#accordion" href="#<%=keys.to_s+"panel"%>" aria-expanded="true" onclick="$(this).find(':first-child').toggleClass('glyphicon glyphicon-minus-sign glyphicon glyphicon-plus-sign')">
+                <span class="glyphicon glyphicon<%= (keys.to_s.include?("Hardware")? "-minus-sign":"-plus-sign")%>" aria-hidden="true"></span>
+                <%=keys%>
```
Missing spaces.

lzap commented on app/views/discovered_hosts/show.html.erb:
```diff
@@ -2,15 +2,46 @@
+            <div class="panel-body">
+              <%@facts.sort_by { |k,v| k.downcase} %>
+              <% @facts.sort.select{ |key, value|  keys.to_s.include?("Miscellaneous") || val.any?{|v| key.include? v  }}.each do |fact| %>
+                <li class="list-group-item"> <strong> <%=fact[0] %>: </strong>  <%= truncate(breaking_wrap_wrap(fact[1],50),:length => 80)%></li>
```
Ditto

lzap commented on app/views/discovered_hosts/show.html.erb:
```diff
@@ -2,15 +2,46 @@
+              </a>
+            </h4>
+          </div>
+          <div id="<%=keys.to_s+"panel"%>" class="panel-collapse collapse <%="in" if keys.to_s.include?"Hardware"%>" role="tabpanel">
```
Ditto.

lzap commented:
Before you start reading this, please discover a host using this helper script: https://github.com/lzap/bin-public/blob/master/discover-host with "real" parameter:

```
 discover-host http://localhost:3000 admin:admin real
```

This gives us some decent fact list from DELL PowerEdge R810. Now to my review:

I think "Storage" should go first and be the only expanded by default. It looks like the "Hardware" section can be easily overhelmed with lots of cpu cores.

Move `hostname` from highlights, its static (never changes). The same for "os" which is always CentOS7.

I still see `ipaddress` in Miscellanaeous section. And I see some `ipmi_` facts in Network section. I think it is worth introducing "IPMI" section with all `ipmi_` facts.

Can you convert `blockdevice_*_size` to human-readable form (XYZ GB)?

Some facts are reported as JSON hashes, why is that @ohadlevy ?

Also can you do the "expansion" javascript for long lines? Or instead of javascript, you can make a link into /fact_values for the particular one to see it complete.

alongoldboim commented:
@lzap I've used the script it was very useful.
-moved Storage to be on top.
-cleared hostname, os from highlights.
-moved ipaddress to highlights, added a ipmi section when needed.
-blockdevice_*_size changed to GB.
-added a tool tip for truncated values.

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -25,8 +26,19 @@ def index
+                   :Network => ["ipaddress", "interfaces", "dhcp", "fqdn", "hostname", "link", "mtu", "net"] + @host.facts_hash["interfaces"].split(","),
+                   :Software => ["bios", "os", "discovery"]}
+    @facts = @host.facts_hash
+    @facts.delete("memorysize_mb")
```
I'd prefer not to delete facts, can you simply move it to Hardware?

alongoldboim replied:
can i make @facts a shallow copy of the facts hash and then the source will not be influenced?

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -25,8 +26,19 @@ def index
+    @facts.delete("memorysize_mb")
+    @facts.each do |key, val|
+       @highlights.merge!({:IPMI => ["ipmi"]}) if key.to_s.downcase.include?("ipmi")
+       @facts[key] = number_to_human_size(val)  if key.to_s.downcase.include?("blockdevice") && key.to_s.downcase.include?("_size")
```
Can you use (case insensitive) regular expressions which are faster than this? Also you can leverage begin/end of the string.

lzap commented on app/helpers/discovered_hosts_helper.rb:
```diff
@@ -65,5 +65,11 @@ def discovery_status_icon(host)
+  def breaking_wrap_wrap(txt, col = 80)
+    txt.gsub(/(.{1,#{col}})( +|$\n?)|(.{1,#{col}})/,
+             "\\1\\3\n")
   end
```
Ok please remove unnecessary wrapping and just truncate long facts appending a "..." there. Make a sane default maximum length so it fits there.

alongoldboim replied:
Ok wanted to give the facts a bit more length without overflowing, but like you said ill just truncate it.

lzap commented:
Ok few comments, it looks good. Can you please add a button between Select Action and Delete "Expand All" with a proper function? No need to do a Collapse All button I think. Then we are RTM!

lzap commented:
LGTM @ohadlevy opinions about this UI?

ohadlevy commented:
can you attach recent UI?

alongoldboim commented:
![selection_006](https://cloud.githubusercontent.com/assets/11405684/9382543/9d1419f8-474a-11e5-8afe-bc7df0130a4d.png)

ohadlevy commented:
I think some of the facts name needs to have pretty names (like
physicalprocesscount) you could consider using table-zebra etc

lzap commented:
Hmm I don't like changing expected names much, zebra is a good idea. Thanks.

alongoldboim commented:
Added zebra to the list, imho we should add assets dir and place scripts and css in an organized way.
Your call :), Thanks for the comments.

lzap commented:
Ok last issues:
- the expand all button has wrong text, should be: "Expand All"
- can you please move all blockdevice_ facts from Highlights back to Storage (as it is pre-expanded and better place to have them)
- I still see ipmi_ facts both in IPMI and Network groups (should only be in IPMI group)

That's it!

alongoldboim commented:
@lzap should be good now :+1:

lzap commented:
Ok looks good, can you reword the commit message? I see some extra text there. Also, why you are not the author? Looks like you messed up with git a bit. This is what I get when I try to commit:

```
Author:     Stephen Benjamin <stbenjam@redhat.com>
AuthorDate: Wed Jul 22 14:36:13 2015 -0400
Commit:     Lukas Zapletal <lzap+git@redhat.com>
CommitDate: Wed Aug 26 13:44:11 2015 +0200

    Fixes 11008 - Changed the view for discovered hosts show page.

    chagne to impi:

    chagne to impi:
```

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -25,8 +26,19 @@ def index
+    @facts = @host.facts_hash.clone
+    @facts.delete("memorysize_mb")
+    @facts.each do |key, val|
+      @categories.merge!({:IPMI => ["ipmi"]}) if key.to_s.downcase.include?("ipmi") && !@categories[:IPMI]
```
Please do `if /^ipmi_/.match(key)` here

lzap commented on app/views/discovered_hosts/show.html.erb:
```diff
@@ -2,15 +2,55 @@
+  $(document).ready(function(){
+    $('.facts-list > li:nth-child(even)').css('background-color','#f5f5f5')
+  });
+</script>
```
Missing newline

lzap commented on app/views/discovered_hosts/show.html.erb:
```diff
@@ -2,15 +2,55 @@
+              <% @facts.sort.select{|key, value| val.any?{|v| (key.include?(v) &&  !key.include?("bios") && !key.include?("ipmi")) || (key.eql? "ipaddress")}}.each do |fact| %>
+                <li class="list-group-item"> <strong> <%= fact[0] %>: </strong>  <%= truncate(fact[1], :length => 50)%></li>
+                <% @facts.delete(fact[0]) %>
+              <% end %>
```
Ok this is code logic in view template, not good. Can you move all the code logic in the controller? Pass the data as hash of hashes and in views just blindly render them.

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -25,8 +26,19 @@ def index
+    @general =    {:Highlights => [ "productname", "memorysize", "interface", "manufacturer", "architecture", "macaddress", "processorcount", "physicalprocessorcount", "discovery_boot"]}
+    @categories = {:Storage => ["blockdevice"],
+                   :Hardware => ["hardw", "manufacturer", "memo", "process"],
+                   :Software => ["bios", "os", "discovery"]}
```
I wonder if you can make this configurable, my idea is that every category could be represented using _single_ regexp. Then you can load these from Settings (create new four settings named `discovery_storage_CATEGORY`) and provide sane default values for them. Example for hardware: `^(hardw|manufacturer|memo|process)".

Please get rid of other extra fact name comarisons in the code. Everything should have been categorized by the four regexps. No other include/match/delete calls in the code.

lzap replied:
In the regexp try to follow the pattern of `^(prefix1|prefix2)` but in case you want to compare by suffixes you can of course do something like `((suffix1|suffix2)$)|(^(prefix1|prefix2))`

lzap replied:
Wait, intstead of sane default, make empty value the default one and if there is no user defined one, use the pre-defined regexp. That's better for future upgrades.

alongoldboim commented:
@lzap applied the changes.

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -24,11 +25,46 @@ def index
+    regex_hightlights = Setting[:discovery_CATEGORY_highlights].empty? ? /^(productname|memorysize|interface|manufacturer|architecture|macaddress$|processorcount|physicalprocessorcount|discovery_boot|ipaddress$)/ : Regexp.new(eval(Setting[:discovery_CATEGORY_highlights]))
+    regex_storage = Setting[:discovery_CATEGORY_storage].empty? ? /^blockdevice/ : Regexp.new(eval(Setting[:discovery_CATEGORY_storage]))
+    regex_hardware = Setting[:discovery_CATEGORY_hardware].empty? ? /^(hardw|manufacturer|memo|process)/ : Regexp.new(eval(Setting[:discovery_CATEGORY_hardware]))
+    regex_network = Setting[:discovery_CATEGORY_network].empty? ? /^(ipaddress|interfaces|dhcp|fqdn|hostname|link|mtu|net)|(#{@host.facts_hash["interfaces"].gsub(",","|")})/ : Regexp.new(eval(Setting[:discovery_CATEGORY_network]))
```
Hmmm I'd rather keep all the regexps clean and static. I think if we copy few facts into highlights AND network, it's no issue. Just rewrite it to be simple regexp without this magic.

lzap replied:
Reading your logic below, it looks like you pick up first, so there must be a way to achieve the same result without generating regexp from interfaces.

alongoldboim replied:
OK, will figure out a way, but just to know whats so bad in generating a regex from the interfaces?

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -24,11 +25,46 @@ def index
+      end
+      @facts.delete(key)
+    end
+    @categories.each {|key,val| @categories.delete(key) if @categories[key].empty? }
```
Why don't you put regexps into an array, so you can iterate over groups. Then you can get rid of lines 41-46 and also get rid of that case statement (iterate through all regexps, if any match then determine group name, put it into the correct group name). Then it will be much shorter.

lzap replied:
Make sure you process group names array in correct order. Putting it into an array should be sufficient.

alongoldboim replied:
OK sounds good.

lzap commented on app/models/setting/discovered.rb:
```diff
@@ -42,6 +48,18 @@ def self.load_defaults
+      ].compact.each { |s| self.create s.update(:category => "Setting::Discovered")}
+    end
+
+
```
I apologize for complications, but I like `discovery_facts_highlights` much better than (capitalized) CATEGORY.

lzap commented:
Ok that looks so much better.

alongoldboim commented:
Thanks it will be done soon :)

alongoldboim commented:
@lzap fixed, there are two versions one with iterating on the array of regex and the other with the case, tell me what commit you prefer, thanks.

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -171,6 +177,27 @@ def auto_provision_all
+    @categories_names = [:Hightlights, :Storage, :Hardware, :Network, :Software, :IPMI]
+  end
+
+  def assign_to_category(key, value )
```
Maybe `assign_fact_to_category` ? Also extra space.

lzap commented on app/helpers/discovered_hosts_helper.rb:
```diff
@@ -10,16 +10,22 @@ def discovered_hosts_title_actions(host)
+                              method = action[2] if action.size > 1
+                              link_to(action[0] , action[1], method)
+                            end.flatten
+      ),
```
Please get rid of code reindenting. This makes cherry picking hard.

lzap commented on app/views/discovered_hosts/show.html.erb:
```diff
@@ -2,15 +2,52 @@
+    $('.facts-list > li:nth-child(even)').css('background-color','#f5f5f5')
+  });
+</script>
+
```
Extra newline :-)

lzap commented:
Ok this looks so much better, now please get rid of duplicating of facts and deleting facts. It is not necessary, you don't use @fact variable anymore in the template. No need to delete anything. Also modify your code in a way that if there is no match found, assign to the "Misceleneous" or "Other" section which is currently missing, but should have been there (no regexp, its a "fallback" group).

alongoldboim commented:
@lzap done.

lzap commented on app/helpers/discovered_hosts_helper.rb:
```diff
@@ -65,5 +71,7 @@ def discovery_status_icon(host)
+
   end
+
 end
```
Please remove all extra reformatting and empty lines.

lzap commented on app/helpers/discovered_hosts_helper.rb:
```diff
@@ -17,9 +17,15 @@ def discovered_hosts_title_actions(host)
+          :class => "btn btn-danger", :confirm => _('Are you sure?'), :method => :delete)
       )
+
     )
```
Same here.

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -171,6 +174,29 @@ def auto_provision_all
+    network = Setting[:discovery_facts_network].empty? ? /^(ipaddress|interfaces|dhcp|fqdn|hostname|link|mtu|net|macaddress|wol|port|speed)/ : Regexp.new(eval(Setting[:discovery_facts_network]))
+    software = Setting[:discovery_facts_software].empty? ? /^(bios|os|discovery)/ : Regexp.new(eval(Setting[:discovery_facts_software]))
+    ipmi = Setting[:discovery_facts_ipmi].empty? ? /^ipmi/ : Regexp.new(eval(Setting[:discovery_facts_ipmi]))
+    @regex_array = [hightlights, storage, hardware, network, software, ipmi, "misceleneous"]
```
Use symbol instead. Alternatively, you can use the index, but I kinda like this solution. Another more clever approach is to create a class named `AlwaysMatcher` with one method `def match; true; end` adding it as the last element. Then the loop can be even more simple (it will always match). But this works for me too.

alongoldboim replied:
changed it a bit, didn't use ur solution but improved the one i had.

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -24,11 +25,13 @@ def index
     @range = nil
-
-    # summary report text
     @report_summary = nil
```
And please keep the comments here, we shoould not remove those important lines.

lzap commented:
Ok almost there, few more glitches and then I will re-test :-)

alongoldboim commented:
@lzap finished, I think we have reached the holy land :+1:

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -29,6 +30,11 @@ def show
     @report_summary = nil
+    init_regex_and_categories
+    @host.facts_hash.each do |key, value|
+      value = number_to_human_size(value)  if /size$/.match(key)
```
Double space before `if`.

Can you also proceed with expanding Settings chapter in our manual? http://theforeman.org/plugins/foreman_discovery/4.0/index.html#3.1.4Globalsettings it is this page: https://github.com/theforeman/theforeman.org/blob/gh-pages/plugins/foreman_discovery/4.0/index.md

alongoldboim commented:
Fixed the double space.
the pr for the documentation part https://github.com/theforeman/theforeman.org/pull/434, should be merged after this pr will be merged.

lzap commented:
Great, thanks. Can only merge after we fix unit tests on develop.

lzap commented:
Ok we can re [test] now.

alongoldboim commented:
Cool, thanks for the reviews @lzap .

lzap commented:
> Cool, thanks for the reviews @lzap .

Ok, I think we can extend this further. Would you mind showing also
interfaces that were discovered?

It's host.interfaces entity, basically in a table with the following
columns:

Primary/Provisioning icons
Identifier
Type
MAC
IP

Very similar to New Host - Interfaces table.

This would be very useful!

## 

Later,
 Lukas #lzap Zapletal

alongoldboim commented:
@lzap Yea we can do that, we already doing that but we are not centering all of the facts together in a table, ill add it, you want to place it under a "Interfaces" category? or place it under an existing one?
Do we want to keep on using our solution with the regex on settings? seen in the pr you referenced that you think about doing it differently.

alongoldboim commented:
Current UX.
![selection_007](https://cloud.githubusercontent.com/assets/11405684/10361349/473afd76-6db0-11e5-8bd1-e14fb1760b31.png)

kybaker commented:
@alongoldboim - You sent a request for me to review. 

The design above looks nice at first glance but it will prove to be cumbersome to use over time. It looks like the goal of this page is to allow users to scan and process data as quickly as possible. Expanding and collapsing adds an extra step to this process. In my experience performing usability testing, users do not want to have to click anything, they want information available to scan as quickly as possible. We are adding an extra step here. I think adding some simple headers above a table for each section would be the most effective user experience.

alongoldboim commented:
@kybaker thanks for the response, i agree with what you'r saying yet there is a problem with expending all the tabs by default because there are lots of facts (can even get to 5 pages of scrolling), that's why we got the highlights section(expanded by default) which should answer most of the user needs and the other facts are organized in categories to help the user get to the needed fact faster by knowing what its connected to.
In the image you can see the scroller size when all is expanded.
![scroller_demo](https://cloud.githubusercontent.com/assets/11405684/10370146/5560196e-6de6-11e5-9b55-f85b9c960cb0.png)

do you still think we should cancel the expanding?

kybaker commented:
I understand and parsing the information here is beneficial. I don't think this will be a show stopper but we could always change if there is negative feedback.

alongoldboim commented:
Sounds good, thanks.

lzap commented:
Yeah I agree with @alongoldboim - we expand the most important stuff anyway, but big servers with hudreds of cores generates _huge_ output. Its like neverending.

@alongoldboim ok your first version looks good, haven't tested it yet but can you make the table padding the same size as in other panes? Also remove the "Type" column and just add icons as we do on the Interfaces tab on the Edit Host form (primary, provisioning) with mouse hoovers. And that should do it!

alongoldboim commented:
@lzap Done, just how i'm supposed to decide whether primary, provisioning flags are active?.

lzap commented:
Ok almost there:
- Can you add 'Back' button that does browser's back action?
- Expand Interfaces pane by default
- Use 'break-me' CSS class I've introduced in Core last week for fact value cells (https://github.com/theforeman/foreman/commit/c687f3a1c9d157b1313070ddc005b518af30acad)
- Let's still use your trimming code, but now for very long strings. I suggest to trim only values longer than 1000 characters, what you think? Or maybe we can get rid of this completely, if it still looks okay with real data.

alongoldboim commented:
Added the requirements, but some of the facts values are placed in 'li' and not in 'td', if I add this class to those elements it doesn't looks so good(I can either change it all to tables or for first version leave it as it is).
About the trimming it works well with all the data I have checked, i think we'd better use the current trimming because it prevents overflow and if we will have issues we could change it.

lzap commented:
Discovered hosts have `.subnet` association, can you please add it as an extra fact to Highlight pane? Would be named "discovery_subnet", you can render subnet name plus the subnet address with a link to the subnet:

discovery_subnet: mysubnet (192.168.44.0)

alongoldboim commented:
Added this fact at the format you wanted.
There is no show page for subnet, i can link it to the edit page but i don't know how valuable it will be, for now i left it as a regular fact without a link, let me know if you think differently.

lzap commented:
Oh right, I think link is not necessary youre right. Testing.

lzap commented:
Ok, I still don't like how trimming works, I think we trim too much on larger screens. Isn't there any other way? Could you talk to Shimon who implemented the "eclipse" CSS style and use that at least? This will only trim if there is no free space. Other than that we are good to go! :-)

alongoldboim commented:
It was Tomer :), added and working, give it a check.

lzap commented:
Ok two last things:

1) N+1 query

```
| N+1 Query detected
 |   Nic::Managed => [:subnet]
 |   Add to your finder: :includes => [:subnet]
 | N+1 Query method call stack
 |   /home/lzap/work/foreman/app/models/host/base.rb:83:in `subnet'
 |   /home/lzap/work/foreman_discovery/app/controllers/discovered_hosts_controller.rb:304:in `add_custom_facts'
 |   /home/lzap/work/foreman_discovery/app/controllers/discovered_hosts_controller.rb:39:in `show'
 |   /home/lzap/work/foreman/app/controllers/concerns/application_shared.rb:13:in `set_timezone'
 |   /home/lzap/work/foreman/app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 |   /home/lzap/work/foreman/lib/middleware/catch_json_parse_errors.rb:9:in `call'
 | 
 | /home/lzap/work/foreman/app/models/host/base.rb:83:in `subnet'
 | /home/lzap/work/foreman_discovery/app/controllers/discovered_hosts_controller.rb:304:in `add_custom_facts'
 | /home/lzap/work/foreman_discovery/app/controllers/discovered_hosts_controller.rb:39:in `show'
 | /home/lzap/work/foreman/app/controllers/concerns/application_shared.rb:13:in `set_timezone'
 | /home/lzap/work/foreman/app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
```

2) The interfaces table lost it's padding, can you put it back?

lzap commented on app/controllers/discovered_hosts_controller.rb:
```diff
@@ -262,4 +293,17 @@ def skip_bullet
+
+  def add_custom_facts
+    if @host.subnet
+      discovery_subnet = "#{@host.subnet.name + " (" + @host.subnet.network + ")"}"
```
Weird usage of embedding :-) I'd probably go with `"#{host.subnet.name} (#{host.subnet.network})"`.

lzap commented:
Actually the pane with no padding is not bad at all, if you can do the same for all that's fine. But I see some problem at the bottom line, there is some unexpected padding still.

![9a9af915-7417-4caf-9901-d3431a42c7c5](https://cloud.githubusercontent.com/assets/49752/11039274/1d705994-8707-11e5-8974-cc30823a8bb3.png)

lzap commented:
The puzzle solution is:

``` diff
diff --git a/app/controllers/discovered_hosts_controller.rb b/app/controllers/discovered_hosts_controller.rb
index 09e2d24..acef4e6 100644
--- a/app/controllers/discovered_hosts_controller.rb
+++ b/app/controllers/discovered_hosts_controller.rb
@@ -244,7 +244,7 @@ class DiscoveredHostsController < ::ApplicationController
     end
   end

-  def find_by_name(includes = "")
+  def find_by_name(*includes)
     params[:id].downcase! if params[:id].present?
     @host = includes.empty? ? resource_base.find_by_id(params[:id]) : resource_base.includes(includes).find_by_id(params[:id])
     @host ||= includes.empty? ? resource_base.find_by_name(params[:id]) : resource_base.includes(includes).find_by_name(params[:id])
@@ -252,7 +252,7 @@ class DiscoveredHostsController < ::ApplicationController
   end

   def find_by_name_incl_subnet
-    find_by_name(:subnet)
+    find_by_name({:interfaces => :subnet})
   end

   def find_multiple
@@ -306,7 +306,7 @@ class DiscoveredHostsController < ::ApplicationController

   def add_custom_facts
     unless @host.primary_interface.subnet.nil?
-      discovery_subnet = @host.primary_interface.subnet.name + "(" + @host.primary_interface.subnet.network + ")"
+      discovery_subnet = "#{@host.primary_interface.subnet.name} (#{@host.primary_interface.subnet.network})"
       assign_fact_to_category("discovery_subnet", discovery_subnet)
     end
   end
```

lzap commented:
Ok good work on this one.

lzap commented on app/models/setting/discovered.rb:
```diff
@@ -42,6 +48,16 @@ def self.load_defaults
+          self.set('discovery_facts_software', N_("The default regex to organize facts for software section"), "" ),
+          self.set('discovery_facts_hardware', N_("The default regex to organize facts for hardware section"), "" ),
+          self.set('discovery_facts_network', N_("The default regex to organize facts for network section"), "" ),
+          self.set('discovery_facts_ipmi', N_("The default regex to organize facts for ipmi section"), "" ),
```
Three more nitpicks:

1) Drop the `The default" from the setting name, I'd go with the short "Regex to organize..."

2) Extra space after `""`

3) Rewrite the commit message, it contains some garbage

lzap replied:
Scratch the (4), we can do separate PR for `""`'s.

lzap commented:
Merged as 8b23c24a667f4e8588f96a346432aafeab63ba85, thank you!
